mathics/core/subexpression.py

```diff
@@ -143,7 +143,7 @@ def to_expression(self):
         parent = self.parent
         p = self.position
         if p == 0:
-            if type(parent) is Symbol:
+            if isinstance(parent, Symbol):
```

Comment:
I used here `type` instead of `isinstance` because it is a little bit faster: `isinstance` checks all the inheritance hierarchy. For this, I exploded that `Symbol` didn't have daughter classes.

Reply:
Please don't do this. Lint checkers will tell you it is wrong. Better again would be to use a property on the type. What is really being asked for is a property like does this class have a leaf?

Reply:
@rocky, I am sure you are right about this. I just was exposing the motivation I had to use this syntax. On the other hand, the "fragility" was also a way to alert about a change in the object model: In the current OM, it does not make any sense to specialize the `Symbol` object. Symbols are just references to `Definition`s.

Reply:
@mmatera I think you are still confused. A WL symbol is composed of an atomic or non decomposable name and a value that is listed under than name. Once the name is defined it can't be changed. Think if it as analogous in Python as defining a function. That function has a name that can't be changed. You can decide never to use that name, or assign have that function name hidden, or assign a variable name initially to that function name but then change it. But the function name for a code object never changes.  And it is in that sense that when we write Symbol("Pi") we what we really mean is the name of the function that was created.

Reply:
> @mmatera I think you are still confused. A WL symbol is composed of an atomic or non decomposable name and a value that is listed under than name.
`Symbol`s do not have a value.  A `Symbol` has a `Definition`, which you access by its name.


> Once the name is defined it can't be changed. Think if it as analogous in Python as defining a function, e.g `def Pi(...)`. That function has a name that can't be changed. You can decide never to use that name, or never use the function, or assign a variable name initially to that function name but then change it. But the function name for a code object never changes. 

A fully agree about this: this is why I have insisted in deal with symbols as singletons.  

> And it is in that sense that when we write Symbol("Pi") we what we really mean is the name of the function that was created. The way a value is associated with that name of course can change.
> 
> And that is why changing `Symbol(Pi)` to `Constant(Pi)` has no effect when function return values of the function change. In both cases you asked using the name `Pi`. There is no way to use a different name but have that _name_ get swapped under the covers to `Pi`.

OK, so the idea is: the difference between `Symbol(Pi)` and `Constant(Pi)` is in the implementation of the `evaluate` method: the second just tries to evaluate if the associated definition has changed, isn't it?

Reply:
> A fully agree about this: this is why I have insisted in deal with symbols as singletons.

What you seem to have done a number of times, is rediscover a familiar concept in programming languages and come up with a possible solution, but not as good as a solution as what has been done before where you to have understood the familiar concept. Woflram modeled parts of Mathematica on Lisp. 

Recall that I started the proliferation of `Symbol(foo)`.  It was my confusion with Lisp Symbol and not realizing how different WL Symbol is and not realizing at the thing that the two should should not be confused. 

What I am suggesting is changing `Constant()` to `SymbolName()`.  While `SymbolName()` is a concept that users see implicitly, and could be a way to describe behavior, it definitely has an aspect and impact about the way we implement expressions.

This kind of thing is analogous to the concept of a "Literal".  For example google "Python literal" and you'll get the idea of some other analogous piece of a programming language that appears and is a concept in the implementation of some language. However in Python there is no "literal" type and you can declare or define something as literal other than by writing the literal value in some way, e.g. 5 or [1, 2, 3],  {"foo":  1}, {"bar"}.

Reply:
Also there should be a separate LiteralValue class which is a parent class for things like Integer, Real, String.

Reply:
> > A fully agree about this: this is why I have insisted in deal with symbols as singletons.
> 
> What you seem to have done a number of times, is rediscover a familiar concept in programming languages and come up with a possible solution, but not as good as a solution as what has been done before were you to have understood the familiar concept. 

Indeed! I am a physicist, not a computer scientist! Sometimes, I realize where to look for implementing a concept, but most of the time, I have to rediscover the wheel to see where is the tireshop :)

> Woflram modeled parts of Mathematica on Lisp.

I know that (barely), but I am not an expert in Lisp. 

> 
> Recall that I started the proliferation of `Symbol(foo)`. It was my confusion with Lisp Symbol and not realizing how different WL Symbol is and not realizing at the time that the two should not be confused, and that was happening with this proliferation.
> 

What is exactly the problem with the `Symbol(foo)` definitions?
 


> What I am suggesting is changing `Constant()` to `SymbolName()`. While `SymbolName()` is a concept that users see implicitly, and could be a way to describe behavior, it definitely has an aspect and impact about the way we implement expressions. A `Symbol` has a `SymbolName` and a `Definition`.
> 

OK, that could work...


> This kind of thing is analogous to the concept of a "Literal". For example google "Python literal" and you'll get the idea of some other analogous piece of a programming language that appears and is a concept in the implementation of some language. However in Python there is no "literal" type and you can declare or define something as literal other than by writing the literal value in some way, e.g. 5 or [1, 2, 3], {"foo": 1}, {"bar"}.

not sure to follow you, but let's try it!

Reply:
> What is exactly the problem with the Symbol(foo) definitions?

One problem of using `Symbol()` when you mean `SymbolName` or Lisp Symbol we have already seen: we now have a circular import, because as a result of pulling in more stuff than is needed in an object.

SymbolNames have a more limited function and use that Symbols.  When you create a large object the encompases everything it leads to undue complexity and it prohibits splitting up the code into smaller more easily testable parts that are more easily understood.  

What would the world be like if we always had to talk about quantum interactions when we refer to biology rather than dividing things into atoms, elements, chemical elements, biology and so on.